File: scripts/database_integration_scripts/enhance_db/src/bin/32_transfer_to_clean_db.rs
/// Transfer all tables from corrupted database to a fresh clean database.
/// Uses ATTACH DATABASE + INSERT INTO ... SELECT for maximum speed (data stays in SQLite engine).
use anyhow::Result;
use rusqlite::Connection;
use std::fs;
use std::io::Write;
use std::time::Instant;

const SRC_DB: &str = "/workspace/data/humans_clean.sqlite3";
const DST_DB: &str = "/workspace/data/humans_clean_new.sqlite3";

fn log(msg: &str) {
    let dur = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap();
    let secs = dur.as_secs();
    let ts = format!(
        "{:02}:{:02}:{:02}",
        (secs % 86400) / 3600,
        (secs % 3600) / 60,
        secs % 60
    );
    let line = format!("[{}] {}", ts, msg);
    println!("{}", line);
    if let Ok(mut f) = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open("/workspace/transfer.log")
    {
        let _ = writeln!(f, "{}", line);
    }
}

fn elapsed(start: Instant) -> String {
    let s = start.elapsed().as_secs();
    if s < 60 {
        format!("{}s", s)
    } else if s < 3600 {
        format!("{}m {}s", s / 60, s % 60)
    } else {
        format!("{}h {}m {}s", s / 3600, (s % 3600) / 60, s % 60)
    }
}

fn main() -> Result<()> {
    let total_start = Instant::now();
    log("=== TRANSFER: Copying all tables to clean database ===");
    log(&format!("Source: {}", SRC_DB));
    log(&format!("Destination: {}", DST_DB));

    // Remove destination if it exists (fresh start)
    if fs::metadata(DST_DB).is_ok() {
        fs::remove_file(DST_DB)?;
        log("Removed existing destination database");
    }

    // Open source database
    let conn = Connection::open(SRC_DB)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL;
         PRAGMA synchronous=NORMAL;
         PRAGMA cache_size=-4000000;
         PRAGMA busy_timeout=60000;
         PRAGMA mmap_size=8589934592;",
    )?;

    // Attach the new clean database
    conn.execute_batch(&format!("ATTACH DATABASE '{}' AS clean;", DST_DB))?;
    conn.execute_batch(
        "PRAGMA clean.journal_mode=WAL;
         PRAGMA clean.synchronous=OFF;
         PRAGMA clean.cache_size=-4000000;",
    )?;
    log("Attached clean database");

    // Get all table creation SQL and index SQL from source
    let mut stmt = conn.prepare(
        "SELECT type, name, sql FROM main.sqlite_master
         WHERE sql IS NOT NULL AND name NOT LIKE 'sqlite_%'
         ORDER BY CASE type WHEN 'table' THEN 0 WHEN 'index' THEN 1 ELSE 2 END, name",
    )?;

    let objects: Vec<(String, String, String)> = stmt
        .query_map([], |row| {
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, String>(1)?,
                row.get::<_, String>(2)?,
            ))
        })?
        .filter_map(|r| r.ok())
        .collect();

    let tables: Vec<&(String, String, String)> =
        objects.iter().filter(|(t, _, _)| t == "table").collect();
    let indexes: Vec<&(String, String, String)> =
        objects.iter().filter(|(t, _, _)| t == "index").collect();

    log(&format!(
        "Found {} tables and {} indexes to transfer",
        tables.len(),
        indexes.len()
    ));

    // Phase 1: Create tables and copy data
    log("--- Phase 1: Creating tables and copying data ---");
    for (_, name, create_sql) in &tables {
        let step_start = Instant::now();

        // Create table in clean database (replace implicit "main." with "clean.")
        let clean_sql = create_sql.replace("CREATE TABLE ", "CREATE TABLE clean.");
        let clean_sql = clean_sql.replace("CREATE TABLE clean.clean.", "CREATE TABLE clean.");
        conn.execute_batch(&clean_sql)?;

        // Copy data
        let copy_sql = format!(
            "INSERT INTO clean.\"{}\" SELECT * FROM main.\"{}\"",
            name, name
        );
        let copied = conn.execute(&copy_sql, [])?;

        log(&format!(
            "  {} : {} rows copied ({})",
            name,
            copied,
            elapsed(step_start)
        ));
    }

    // Phase 2: Create indexes
    log("--- Phase 2: Creating indexes ---");
    for (_, name, create_sql) in &indexes {
        let step_start = Instant::now();

        // Rewrite index SQL to target clean database
        let clean_sql = create_sql
            .replace("CREATE INDEX ", "CREATE INDEX clean.")
            .replace("CREATE UNIQUE INDEX ", "CREATE UNIQUE INDEX clean.");
        // Fix double-prefixed clean.clean.
        let clean_sql = clean_sql
            .replace("clean.clean.", "clean.");

        match conn.execute_batch(&clean_sql) {
            Ok(_) => log(&format!("  Index {} created ({})", name, elapsed(step_start))),
            Err(e) => log(&format!("  Index {} FAILED: {} ({})", name, e, elapsed(step_start))),
        }
    }

    // Phase 3: Verify
    log("--- Phase 3: Verification ---");
    for (_, name, _) in &tables {
        let src_count: i64 = conn
            .query_row(
                &format!("SELECT COUNT(*) FROM main.\"{}\"", name),
                [],
                |r| r.get(0),
            )
            .unwrap_or(-1);
        let dst_count: i64 = conn
            .query_row(
                &format!("SELECT COUNT(*) FROM clean.\"{}\"", name),
                [],
                |r| r.get(0),
            )
            .unwrap_or(-1);
        let status = if src_count == dst_count { "OK" } else { "MISMATCH!" };
        log(&format!(
            "  {} : src={} dst={} [{}]",
            name, src_count, dst_count, status
        ));
    }

    // Detach and set final pragmas on new DB
    conn.execute_batch("DETACH DATABASE clean;")?;

    // Open the new DB directly to finalize
    let new_conn = Connection::open(DST_DB)?;
    new_conn.execute_batch("PRAGMA journal_mode=WAL; PRAGMA integrity_check;")?;
    let integrity: String = new_conn.query_row("PRAGMA integrity_check;", [], |r| r.get(0))?;
    log(&format!("New database integrity: {}", integrity));

    log(&format!(
        "=== TRANSFER COMPLETE ({}) ===",
        elapsed(total_start)
    ));
    log(&format!("Clean database at: {}", DST_DB));

    Ok(())
}

File: scripts/database_integration_scripts/legacy/32_transfer_to_clean_db.py
"""32 - Transfer all tables from source DB to a fresh clean target DB.

Mirrors `enhance_db/src/bin/32_transfer_to_clean_db.rs`.

  Inputs : a corrupted/working SQLite DB (default: data/humans_clean.sqlite3)
  Output : a fresh DB with the same schema and data
           (default: data/humans_clean_new.sqlite3)

Strategy
--------
ATTACH DATABASE 'dst' AS clean, then copy each table via
INSERT INTO clean.<t> SELECT * FROM main.<t>. Skips sqlite internal tables.

Usage
-----
    python3 32_transfer_to_clean_db.py            # synthetic DB
    python3 32_transfer_to_clean_db.py --full     # uses default paths
"""

from __future__ import annotations

import os
import sqlite3
import tempfile
from pathlib import Path

from common import DB_PATH, insert_rows, log, parse_run_mode


def run(src_path: Path | str, dst_path: Path | str) -> dict[str, int]:
    src_path = str(src_path)
    dst_path = str(dst_path)
    log(f"[DB] 32: transfer {src_path} -> {dst_path}")

    if os.path.exists(dst_path):
        os.remove(dst_path)

    conn = sqlite3.connect(src_path)
    conn.execute("PRAGMA journal_mode=WAL")
    conn.execute("PRAGMA synchronous=NORMAL")
    conn.execute(f"ATTACH DATABASE '{dst_path}' AS clean")
    conn.execute("PRAGMA clean.journal_mode=WAL")
    conn.execute("PRAGMA clean.synchronous=OFF")

    objects = conn.execute(
        "SELECT type, name, sql FROM main.sqlite_master "
        "WHERE sql IS NOT NULL AND name NOT LIKE 'sqlite_%' "
        "ORDER BY CASE type WHEN 'table' THEN 0 WHEN 'index' THEN 1 ELSE 2 END, name"
    ).fetchall()
    tables = [o for o in objects if o[0] == "table"]
    indexes = [o for o in objects if o[0] == "index"]
    log(f"[32] Found {len(tables)} tables and {len(indexes)} indexes")

    counts: dict[str, int] = {}
    for _, name, sql in tables:
        clean_sql = sql.replace("CREATE TABLE ", "CREATE TABLE clean.", 1)
        clean_sql = clean_sql.replace("CREATE TABLE clean.clean.", "CREATE TABLE clean.")
        conn.execute(clean_sql)
        cur = conn.execute(f'INSERT INTO clean."{name}" SELECT * FROM main."{name}"')
        counts[name] = cur.rowcount
        log(f"[32]   {name}: {cur.rowcount} rows copied")

    for _, name, sql in indexes:
        clean_sql = (
            sql.replace("CREATE INDEX ", "CREATE INDEX clean.")
               .replace("CREATE UNIQUE INDEX ", "CREATE UNIQUE INDEX clean.")
        )
        clean_sql = clean_sql.replace("clean.clean.", "clean.")
        try:
            conn.execute(clean_sql)
        except sqlite3.Error as e:
            log(f"[32]   index {name} FAILED: {e}")

    conn.commit()

    for _, name, _ in tables:
        s = conn.execute(f'SELECT COUNT(*) FROM main."{name}"').fetchone()[0]
        d = conn.execute(f'SELECT COUNT(*) FROM clean."{name}"').fetchone()[0]
        status = "OK" if s == d else "MISMATCH"
        log(f"[32]   verify {name}: src={s} dst={d} [{status}]")

    conn.execute("DETACH DATABASE clean")
    conn.close()

    new_conn = sqlite3.connect(dst_path)
    integrity = new_conn.execute("PRAGMA integrity_check").fetchone()[0]
    new_conn.close()
    log(f"[32] integrity: {integrity}")
    return counts


def _sample_main() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        src = Path(tmp) / "src.sqlite3"
        dst = Path(tmp) / "dst.sqlite3"
        with sqlite3.connect(src) as seed:
            seed.execute("CREATE TABLE foo (id INTEGER PRIMARY KEY, label TEXT)")
            seed.execute("CREATE TABLE bar (k TEXT, v INTEGER)")
            seed.execute("CREATE INDEX idx_foo_label ON foo(label)")
            insert_rows(seed, "foo", [{"id": 1, "label": "alpha"}, {"id": 2, "label": "beta"}])
            insert_rows(seed, "bar", [{"k": "x", "v": 10}, {"k": "y", "v": 20}, {"k": "z", "v": 30}])

        counts = run(src, dst)
        with sqlite3.connect(dst) as check:
            foo_rows = check.execute("SELECT * FROM foo").fetchall()
            bar_rows = check.execute("SELECT * FROM bar").fetchall()
        log(f"[sample] copied counts={counts}")
        log(f"  foo: {foo_rows}")
        log(f"  bar: {bar_rows}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        run(DB_PATH, DB_PATH.parent / "humans_clean_new.sqlite3")
    else:
        _sample_main()
